### PR TITLE
perf(agg): add benchmark for foldable aggregator

### PR DESCRIPTION
**Pull request by Honeta**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

as title

## Checklist For Contributors

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

benchmark
```
test bench_foldable_count_agg_vis_rate_0_05 ... bench:     189,442 ns/iter (+/- 1,419)
test bench_foldable_count_agg_vis_rate_0_25 ... bench:     209,099 ns/iter (+/- 1,080)
test bench_foldable_count_agg_vis_rate_0_5  ... bench:     236,969 ns/iter (+/- 614)
test bench_foldable_count_agg_vis_rate_0_75 ... bench:     271,424 ns/iter (+/- 2,308)
test bench_foldable_count_agg_without_vis   ... bench:     129,485 ns/iter (+/- 477)
test bench_foldable_max_agg_vis_rate_0_05   ... bench:     219,781 ns/iter (+/- 2,141)
test bench_foldable_max_agg_vis_rate_0_25   ... bench:     232,116 ns/iter (+/- 4,050)
test bench_foldable_max_agg_vis_rate_0_5    ... bench:     267,885 ns/iter (+/- 3,006)
test bench_foldable_max_agg_vis_rate_0_75   ... bench:     288,743 ns/iter (+/- 693)
test bench_foldable_max_agg_without_vis     ... bench:     257,916 ns/iter (+/- 448)
test bench_foldable_min_agg_vis_rate_0_05   ... bench:     220,225 ns/iter (+/- 2,015)
test bench_foldable_min_agg_vis_rate_0_25   ... bench:     235,350 ns/iter (+/- 1,213)
test bench_foldable_min_agg_vis_rate_0_5    ... bench:     261,232 ns/iter (+/- 1,756)
test bench_foldable_min_agg_vis_rate_0_75   ... bench:     289,783 ns/iter (+/- 967)
test bench_foldable_min_agg_without_vis     ... bench:     229,307 ns/iter (+/- 356)
test bench_foldable_sum_agg_vis_rate_0_05   ... bench:     235,178 ns/iter (+/- 1,452)
test bench_foldable_sum_agg_vis_rate_0_25   ... bench:     301,981 ns/iter (+/- 972)
test bench_foldable_sum_agg_vis_rate_0_5    ... bench:     405,590 ns/iter (+/- 3,253)
test bench_foldable_sum_agg_vis_rate_0_75   ... bench:     538,044 ns/iter (+/- 5,959)
test bench_foldable_sum_agg_without_vis     ... bench:     609,232 ns/iter (+/- 6,964)
```

### CODE REVIEW

File: src/stream/src/executor/aggregation/agg_impl/foldable.rs [hunk 1]
```diff
@@ -429,9 +429,17 @@ impl_fold_agg! { F64Array, Float64, F32Array }
...
+
+    use itertools::Itertools;
+    use rand::seq::SliceRandom;
+    use rand::Rng;
+    use risingwave_common::array::stream_chunk::Op;
+    use risingwave_common::array::{Array, ArrayBuilder, I64Array, I64ArrayBuilder};
+    use risingwave_common::buffer::{Bitmap, BitmapBuilder};
     use risingwave_common::types::F64;
     use risingwave_common::{array, array_nonnull};
+    use test::Bencher;
```
Comment (kwannoel) [inline src/stream/src/executor/aggregation/agg_impl/foldable.rs:442]: I think we should use `Criterion` for benchmarking, it is more stable than `bencher`.

Comment (kwannoel) [reply src/stream/src/executor/aggregation/agg_impl/foldable.rs:442]: that being said, if it's too much trouble to migrate, and bencher serves our purposes here (results are accurate and consistent), feel free to ignore this comment.

Comment (Honeta) [reply src/stream/src/executor/aggregation/agg_impl/foldable.rs:442]: I've tried using Criterion for benchmarking, but the excellent encapsulation of the test object failed my attempt :(

File: src/stream/src/executor/aggregation/agg_impl/foldable.rs [hunk 2]
```diff
@@ -649,4 +657,354 @@ mod tests {
...
+                            Op::Delete
+                        };
+                        ops.push(op);
+                        if op == Op::Insert {
+                            let value = rng.gen::<i32>() as i64;
+                            data_builder.append_n(1, Some(value));
+                            cur_data.push(value);
+                        } else {
+                            let idx = rng.gen_range(0..cur_data.len());
+                            data_builder.append_n(1, Some(cur_data[idx]));
```
Comment (kwannoel) [inline src/stream/src/executor/aggregation/agg_impl/foldable.rs:715]:
```suggestion
                            let value = rng.gen::<i32>() as i64;
                            data_builder.append(Some(value));
                            cur_data.push(value);
                        } else {
                            let idx = rng.gen_range(0..cur_data.len());
                            data_builder.append(Some(cur_data[idx]));
```

File: src/stream/src/executor/aggregation/agg_impl/foldable.rs [hunk 3]
```diff
@@ -649,4 +657,354 @@ mod tests {
...
+        if let Some(bitmap) = bitmap.as_ref() {
+            for i in 0..chunk_size {
+                // SAFETY(value_at_unchecked): the idx is always in bound.
+                unsafe {
+                    if bitmap.is_set_unchecked(i) {
+                        let op = if append_only || cur_data.is_empty() || rng.gen() {
+                            Op::Insert
+                        } else {
+                            Op::Delete
+                        };
```
Comment (kwannoel) [inline src/stream/src/executor/aggregation/agg_impl/foldable.rs:707]: Can you refactor this out into its own function? Seems like logic is duplicated below.

File: src/stream/src/executor/aggregation/agg_impl/foldable.rs [hunk 4]
```diff
@@ -649,4 +657,354 @@ mod tests {
...
+                    let value = rng.gen::<i32>() as i64;
+                    data_builder.append_n(1, Some(value));
+                    cur_data.push(value);
+                } else {
+                    let idx = rng.gen_range(0..cur_data.len());
+                    data_builder.append_n(1, Some(cur_data[idx]));
+                    cur_data.remove(idx);
+                }
+            }
+        }
```
Comment (kwannoel) [inline src/stream/src/executor/aggregation/agg_impl/foldable.rs:742]: Seems like most of this is duplicated with above, could you please refactor this into a `build_data_chunk` function or something similar?

After that, this should become a lot more readable.

File: src/stream/src/executor/aggregation/agg_impl/foldable.rs [hunk 5]
```diff
@@ -649,4 +657,354 @@ mod tests {
...
+        bench_i64(
+            b,
+            Box::<TestStreamingMaxAgg<I64Array>>::default(),
+            "max",
+            1024,
+            0.05,
+            100,
+            true,
+        );
+    }
```
Comment (kwannoel) [inline src/stream/src/executor/aggregation/agg_impl/foldable.rs:1009]: I think this is fine for now, but you could use a macro or function to generate all these benchmark cases to reduce LoC. ~250 lines are boilerplate.

**Review by kwannoel — COMMENTED** [review]
Generally LGTM, thanks for much for this!

File: src/stream/src/executor/aggregation/agg_impl/foldable.rs (hunk 1 above)
Comment (st1page) [reply src/stream/src/executor/aggregation/agg_impl/foldable.rs:442]: FYI: https://github.com/bheisler/criterion.rs/issues/158

**Review by st1page — COMMENTED** [review]
Sorry for the late review. You can come back to see it next week when you start work. 

I think we can move the logic of the "generate a legal stream chunk" in the `src/common/src/test_utils`

**Review by st1page — APPROVED** [review]
LGTM

**Pull request merged**